OsmAnd-java/src/main/java/net/osmand/router/RouteColorize.java

```diff
@@ -0,0 +1,416 @@
+                slopes[i] = Double.NaN;
+            } else {
+                double[] arg = findDerivativeArguments(distances, elevations, i, slopeRange);
+                slopes[i] = (arg[1] - arg[0]) / (arg[3] - arg[2]);
+            }
+        }
+        return slopes;
+    }
+
+    public List<Data> getResult(boolean simplify) {
+        List<Data> result = new ArrayList<>();
+        if (simplify) {
+            result = simplify();
+        } else {
+            for (int i = 0; i < latitudes.length; i++) {
+                result.add(new Data(i, latitudes[i], longitudes[i], values[i]));
+            }
+        }
+        for (Data data : result) {
+            data.color = getColorByValue(data.val);
+        }
+        return result;
+    }
+
+    public Color getColorByValue(double value) {
+        if (Double.isNaN(value)) {
+            value = (minValue + maxValue) / 2;
+        }
+        for (int i = 0; i < palette.length - 1; i++) {
+            if (value == palette[i][VALUE_INDEX])
+                return new Color((int) palette[i][DECIMAL_COLOR_INDEX]);
+            if (value >= palette[i][VALUE_INDEX] && value <= palette[i + 1][VALUE_INDEX]) {
+                Color minPaletteColor = new Color((int) palette[i][DECIMAL_COLOR_INDEX]);
+                Color maxPaletteColor = new Color((int) palette[i + 1][DECIMAL_COLOR_INDEX]);
+                double minPaletteValue = palette[i][VALUE_INDEX];
+                double maxPaletteValue = palette[i + 1][VALUE_INDEX];
+                double percent = (value - minPaletteValue) / (maxPaletteValue - minPaletteValue);
+                double resultRed = minPaletteColor.getRed() + percent * (maxPaletteColor.getRed() - minPaletteColor.getRed());
+                double resultGreen = minPaletteColor.getGreen() + percent * (maxPaletteColor.getGreen() - minPaletteColor.getGreen());
+                double resultBlue = minPaletteColor.getBlue() + percent * (maxPaletteColor.getBlue() - minPaletteColor.getBlue());
+                double resultAlpha = minPaletteColor.getAlpha() + percent * (maxPaletteColor.getAlpha() - minPaletteColor.getAlpha());
+                return new Color((int) resultRed, (int) resultGreen, (int) resultBlue, (int) resultAlpha);
+            }
+        }
+        return getDefaultColor();
+    }
+
+    public void setPalette(double[][] palette) {
+        this.palette = palette;
+        checkPalette();
+        sortPalette();
+    }
+
+    private Color getDefaultColor() {
+        return new Color(0, 0, 0, 0);
+    }
+
+    private List<Data> simplify() {
+        if (dataList == null) {
+            dataList = new ArrayList<>();
+            for (int i = 0; i < latitudes.length; i++) {
+                //System.out.println(latitudes[i] + " " + longitudes[i] + " " + values[i]);
+                dataList.add(new Data(i, latitudes[i], longitudes[i], values[i]));
+            }
+        }
+        List<Node> nodes = new ArrayList<>();
```

Comment:
Probably it's easier to internally Data extends Node, to avoid conversion for simplification algorithm

Reply:
Error in cast List<Data> to List<Node>